libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,312 @@
+struct __is_move_only_function : false_type {};
+
+template <class _ReturnT, class... _ArgTypes>
+struct __is_move_only_function<move_only_function<_ReturnT(_ArgTypes...)>> : true_type {};
+
+_LIBCPP_END_NAMESPACE_STD
+
+// This header is only partially guarded on purpose. This header is an implementation detail of move_only_function.h
+// and generates multiple versions of std::move_only_function
+#endif // _LIBCPP___FUNCTIONAL_MOVE_ONLY_FUNCTION_IMPL_H
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_CV
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV&
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+#endif
+
+#ifndef _LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT false
+#endif
+
+#define _LIBCPP_MOVE_ONLY_FUNCTION_CVREF _LIBCPP_MOVE_ONLY_FUNCTION_CV _LIBCPP_MOVE_ONLY_FUNCTION_REF
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#ifdef _LIBCPP_ABI_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI [[_Clang::__trivial_abi__]]
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#endif
+
+template <class _ReturnT, class... _ArgTypes>
+class _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI move_only_function<_ReturnT(
+    _ArgTypes...) _LIBCPP_MOVE_ONLY_FUNCTION_CVREF noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT)> {
+  template <class _Functor>
+  struct __function_wrappers_impl {
+    __function_wrappers_impl()                                = delete;
+    __function_wrappers_impl(const __function_wrappers_impl&) = delete;
+    __function_wrappers_impl(__function_wrappers_impl&&)      = delete;
+
+    _LIBCPP_HIDE_FROM_ABI static void __destroy(void* __store) noexcept {
+      if constexpr (!__fits_in_buffer<_Functor>) {
+        std::destroy_at(*static_cast<_Functor**>(__store));
+        ::operator delete(*static_cast<void**>(__store));
+      } else {
+        std::destroy_at(static_cast<_Functor*>(__store));
+      }
+    }
+
+    _LIBCPP_HIDE_FROM_ABI static _ReturnT
+    __call(_LIBCPP_MOVE_ONLY_FUNCTION_CV std::byte* __functor, _ArgTypes... __args) {
+      if constexpr (!__fits_in_buffer<_Functor>) {
+        return std::invoke(static_cast<_Functor _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS>(
+                               **reinterpret_cast<_Functor * _LIBCPP_MOVE_ONLY_FUNCTION_CV*>(__functor)),
+                           std::forward<_ArgTypes>(__args)...);
+      } else {
+        return std::invoke(static_cast<_Functor _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS>(
+                               *reinterpret_cast<_LIBCPP_MOVE_ONLY_FUNCTION_CV _Functor*>(__functor)),
+                           std::forward<_ArgTypes>(__args)...);
+      }
+    }
+  };
+
+  template <class _Functor>
+  using __function_wrappers = __function_wrappers_impl<remove_cvref_t<_Functor>>;
```

Comment:
In what case would it make a difference to use `decay_t` instead of `remove_cvref_t`? I'm not that familiar with `decay_t`, so my default is `remove_cvref_t`.

Reply:
`decay_t<T>` is the type that would be deduced by type deduction (e.g. `auto`); it's different from `remove_cvref_t<T>` specifically for (cvref-qualified) functions and (cvref-qualified) arrays. For example `remove_cvref_t<void(&)()>` is `void()` but `decay_t<void(&)()>` is `void(*)()`. The latter is what we actually want to store.
In this particular source file, I would expect that `remove_cvref_t` is never necessary; it'll always be either equivalent to `decay_t`, or incorrect.